File: peace/EAST/Java/src/eSTAssembly/ESTAssembly.java
package eSTAssembly;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Properties;

import com.mhhe.clrs2e.*;

public class ESTAssembly {
	protected final int INT_MAX = Integer.MAX_VALUE;
	protected final int INT_MIN = Integer.MIN_VALUE;
	Properties props;
	Graph g;	//graph to store all the ests. It is generated in 'readEstFile' function.
	SixTuplesGeneration gen;
	Reconstruction rec;
	InclusionNodes incNodes;

	
	public ESTAssembly(Properties p) {
		props = p;
		incNodes = new InclusionNodes();
		g = new Graph(props, incNodes);
		gen = null;
		rec = null;
	}
	
	public void assemble() {
		readEstFile(props.getProperty("EstFile"));
		readMST(props.getProperty("MSTFile"));
		gen = new SixTuplesGeneration(props, g, incNodes);
		rec = new Reconstruction(props, g, gen.getAlignArray(), gen.getLeftEnds(), incNodes);
		rec.getConsensus();
		System.out.println("number of calls for SW alignment: " + rec.alignment.numCall);
		System.out.println("used time for SW alignment: " + rec.alignment.usedTime);
		System.out.println("number of calls for NW score: " + rec.g.ovl.alignment.numCall2);
		System.out.println("used time for NW score: " + rec.g.ovl.alignment.usedTime2);
	}
	
	/*
	 * Read ests from the input file;
	 * Generate a Graph object, all the ests are considered to be one node in the graph;
	 * No edge in the graph. Edges will be added in "createAlignArray" function.
	 * 
	 * FASTA format:
	 * A sequence in FASTA format begins with a single-line description, followed by lines of 
	 * sequence data. The description line is distinguished from the sequence data by a greater-than
	 * (">") symbol in the first column. The word following the ">" symbol is the identifier of the 
	 * sequence, and the rest of the line is the description (both are optional). There should be no 
	 * space between the ">" and the first letter of the identifier. It is recommended that all lines 
	 * of text be shorter than 80 characters. The sequence ends if another line starting with a ">" 
	 * appears; this indicates the start of another sequence.
	 * 
	 * EST file comment format:
	 * >g001_001169_001679: first one is number of gene, second is index of starting position, third is 
	 * 						index of ending position. Index starts from 0.
	 */
	private void readEstFile(String inFileName) {
		boolean bExists = false;
		ArrayList<String> ests;	//store all the ests. 
		ests = new ArrayList<String> ();
		
		try{ 
			File f = (new File(inFileName));
			bExists = f.exists();
			if (!bExists) {
				System.out.println("File does not exist!");
				return;
			}

			BufferedReader in = new BufferedReader(new FileReader(f));
			String str = in.readLine();
			while (str != null) {
				str = str.trim();
				// first line is comment line which begins from '>'
				if (str.charAt(0) == '>') {	//comment line begins from '>'
					String[] paras = str.split("_");
					ests.add(paras[1]);
					//ests.add("0"); //starting position
					ests.add(str.substring(1)); //comment
					
					//get est in the next lines
					str = in.readLine();
					StringBuffer estStr = new StringBuffer();
					while (str != null) {
						str = str.trim();
						if (str.compareTo("") != 0)	{
							if (str.charAt(0) != '>') {
								estStr.append(str.trim());
							} else  {
								ests.add(estStr.toString());
								break;
							}
						}
						str = in.readLine();
					}
					if (str == null) {
						ests.add(estStr.toString());
					}
				} 
			}
			in.close();			
		}catch(IOException e){ 
			System.out.println(e.toString());
			return;
		}
		
		//generate a graph from the input file
		int i=0;
		while (i<ests.size()) {
			//the sequence is upper-case
			g.addNode(new Node(ests.get(i), ests.get(i+1), ests.get(i+2).toUpperCase()));
			i = i+3;
		}
	}
	
	/*
	 * read a minimum spanning tree from the input MST file
	 */
	private void readMST(String inFileName) {
		int nOfNodes = g.graphNodes.size();
		int[][] nodes = new int[nOfNodes-1][3];	//store edges in MST, there are n-1 edges, n is number of nodes.

		//read mst from the input file
		boolean bExists = false;
		try{ 
			File f = (new File(inFileName));
			bExists = f.exists();
			if (!bExists) {
				System.out.println("File does not exist!");
				return;
			}

			BufferedReader in = new BufferedReader(new FileReader(f));
			String str = in.readLine();
			int curIndex = 0;
			while (str != null) {
				str = str.trim();
				if (str.charAt(0) != '#') {	//comment line begins from '#'
					String[] paras = str.split(",");
					if (Integer.parseInt(paras[0]) != -1) {	//-1 means root of MST
						int i0 = Integer.parseInt(paras[0]);
						int i1 = Integer.parseInt(paras[1]);
						int i2 = Integer.parseInt(paras[2]);
						nodes[curIndex][0] = i0;
						nodes[curIndex][1] = i1;
						nodes[curIndex][2] = i2;
						curIndex++;
					}
				} 
				str = in.readLine();
			}
			in.close();			
		}catch(IOException e){ 
			System.out.println(e.toString());
			return;
		}
		
		// Make a undirected MST.
		WeightedAdjacencyListGraph mst = 
			new WeightedAdjacencyListGraph(nOfNodes, false);
		for (int i=0; i<nOfNodes; i++) {	//i is the index of the node in graph
			mst.addVertex(i, Integer.toString(i));
		}
		for (int j=0; j<nodes.length; j++) {
			mst.addEdge(nodes[j][0], nodes[j][1], nodes[j][2]);
		}
		
		g.setMst(mst);
	}

	protected static Properties getProperties(String fName) throws IOException {
		Properties props = new Properties();
		File f = new File(fName);
        
        if (!f.exists()) {
        	return props;
        }
        
        props.load(new FileInputStream(f)); 
        return props;
    }

	public static void main(String[] args) {
		Properties props = null;
		try {
			if (args.length > 0) {
				props = getProperties(args[0]);
				props.setProperty("EstFile", args[1]);
				props.setProperty("MSTFile", args[2]);
				props.setProperty("ConsensusFile", args[3]);
				props.setProperty("SingletonFile", args[4]);
				props.setProperty("NumOfUsedESTs", args[5]);
			} else {
				props = getProperties("config.properties");
			}
		} catch (IOException e) {
			System.err.println("Get config.properties failed, " + e);
	    	return;
		}

		ESTAssembly assemble = new ESTAssembly(props);

		assemble.assemble();
	}
}

File: peace/EAST/C++/ESTAssembly.cpp
#include "ESTAssembly.h"
#include <cstdlib>
using namespace std;


ESTAssembly::ESTAssembly(std::string& estF, std::string& mstF) {
	incNodes = new InclusionNodes();
	g = new Graph(incNodes);
    gen = NULL;
    rec = NULL;
    estFileName = estF;
    mstFileName = mstF;
}

ESTAssembly::ESTAssembly(std::string& estF, std::string& mstF, std::string& qualF) {
	incNodes = new InclusionNodes();
	g = new Graph(incNodes);
    gen = NULL;
    rec = NULL;
    estFileName = estF;
    mstFileName = mstF;
    qualFileName = qualF;
}

void ESTAssembly::assemble(std::string& con, std::string& sing, std::string& numF) {
	int longestEstLen = 0;
	if (USE_QUALITY_FILE == 1) { //use quality file
		longestEstLen = readEstQualFile();
	} else {
		longestEstLen = readEstFile();
	}
	readMST();
	gen = new SixTuplesGeneration(g, incNodes);
	rec = new Reconstruction(g, gen->getAlignArray(), gen->getLeftEnds(), incNodes, con, sing, numF, longestEstLen);
	rec->getConsensus();
}

ESTAssembly::~ESTAssembly() {
	delete incNodes;
	delete g;
	delete gen;
	delete rec;
}
/*
 * Read ests from the input file;
 * Generate a Graph object, all the ests are considered to be one node in the graph;
 * No edge in the graph. Edges will be added in "createAlignArray" function.
 *
 * Return the length of the longest EST which will be used as COMPARISON_LENGTH in reconstruction.
 *
 * FASTA format:
 * A sequence in FASTA format begins with a single-line description, followed by lines of
 * sequence data. The description line is distinguished from the sequence data by a greater-than
 * (">") symbol in the first column. The word following the ">" symbol is the identifier of the
 * sequence, and the rest of the line is the description (both are optional). There should be no
 * space between the ">" and the first letter of the identifier. It is recommended that all lines
 * of text be shorter than 80 characters. The sequence ends if another line starting with a ">"
 * appears; this indicates the start of another sequence.
 *
 * EST file comment format:
 * >g001_001169_001679: first one is number of gene, second is index of tarting position, third is
 * 						index of ending position. Index starts from 0.
 */
int ESTAssembly::readEstFile() {
	vector<string> ests; //store all the ests.
	int longestLen = 0;

	ifstream in;
	in.open(estFileName.c_str(), ios::in);

	if (in.good()) {
		string str;
		getline(in, str);
		while (!in.eof()) {
			str.erase(str.find_last_not_of(" \n\r\t")+1); //trim str
			if (str.size()==0) getline(in, str);
			// first line is comment line which begins from '>'
			if (str[0] == '>') {	//comment line begins from '>'
				//vector<string> paras = split(str, '_');
				//ests.push_back(paras[1]);
				ests.push_back("0");
				ests.push_back(str.substr(1)); //comment

				//get est in the next lines
				getline(in, str);
				string estStr;
				while (str.size() != 0) {
					str.erase(str.find_last_not_of(" \n\r\t")+1); //trim str
					if (str.size() != 0)	{
						if (str[0] != '>') {
							estStr.append(str);
						} else  {
							ests.push_back(estStr);
							int theSize = estStr.size();
							if (theSize > longestLen)
								longestLen = theSize;
							break;
						}
					}
					getline(in, str);
				}
				if (str.size() == 0) {
					int theSize = estStr.size();
					if (theSize > longestLen)
						longestLen = theSize;
					ests.push_back(estStr);
				}
			}
		}
	}
	in.close();

	//generate a graph from the input file
	int i=0;
	while (i<ests.size()) {
		//the sequence is upper-case
		g->addNode(Node(ests[i], ests[i+1], toUpperCase(ests[i+2])));
		i = i+3;
	}

	return longestLen;
}

int ESTAssembly::readEstQualFile() {
	vector<string> ests; //store all the ests.
	vector<string> scores; //store all the scores corresponding the the ests.
	int longestLen = 0;

	ifstream in;

	//read est file
	in.open(estFileName.c_str(), ios::in);

	if (in.good()) {
		string str;
		getline(in, str);
		while (!in.eof()) {
			str.erase(str.find_last_not_of(" \n\r\t")+1); //trim str
			if (str.size()==0) getline(in, str);
			// first line is comment line which begins from '>'
			if (str[0] == '>') {	//comment line begins from '>'
				//vector<string> paras = split(str, '_');
				//ests.push_back(paras[1]);
				ests.push_back("0");
				ests.push_back(str.substr(1)); //comment

				//get est in the next lines
				getline(in, str);
				string estStr;
				while (str.size() != 0) {
					str.erase(str.find_last_not_of(" \n\r\t")+1); //trim str
					if (str.size() != 0)	{
						if (str[0] != '>') {
							estStr.append(str);
						} else  {
							ests.push_back(estStr);
							int theSize = estStr.size();
							if (theSize > longestLen)
								longestLen = theSize;
							break;
						}
					}
					getline(in, str);
				}
				if (str.size() == 0) {
					int theSize = estStr.size();
					if (theSize > longestLen)
						longestLen = theSize;
					ests.push_back(estStr);
				}
			}
		}
	}
	in.close();

	ifstream in2;
	//read quality file
	in2.open(qualFileName.c_str(), ios::in);

	if (in2.good()) {
		string str;
		getline(in2, str);

		while (!in2.eof()) {
			str.erase(str.find_last_not_of(" \n\r\t")+1); //trim str
			if (str.size()==0) getline(in2, str);
			// first line is comment line which begins from '>'
			if (str[0] == '>') {	//comment line begins from '>'
				//get est in the next lines
				getline(in2, str);
				string estStr;
				while (str.size() != 0) {
					str.erase(str.find_last_not_of(" \n\r\t")+1); //trim str
					if (str.size() != 0)	{
						if (str[0] != '>') {
							estStr.append(str+" ");
						} else  {
							scores.push_back(estStr);
							break;
						}
					}
					getline(in2, str);
				}
				if (str.size() == 0) {
					scores.push_back(estStr);
				}
			}
		}
	}
	in2.close();

	//generate a graph from the input files
	int i=0;
	while (i<ests.size()) {
		//the sequence is upper-case
		g->addNode(Node(ests[i], ests[i+1], toUpperCase(ests[i+2]), getIntScores(scores[i/3])));
		i = i+3;
	}
	return longestLen;
}

//given a record from a quality file, transform it to int scores. The scores in the input string are separated by one space.
std::vector<int> ESTAssembly::getIntScores(const std::string& str) {
	vector<int> ret;
	string tmp = "";
	for (int i=0; i<str.size(); i++) {
		if (str[i] != ' ') {
			tmp.push_back(str[i]);
		} else {
			ret.push_back(atoi(tmp.c_str()));
			tmp = "";
		}
	}
	if (tmp.size() > 0) {
		ret.push_back(atoi(tmp.c_str()));
	}

	return ret;
}

/*
 * read a minimum spanning tree from the input MST file
 */
void ESTAssembly::readMST() {
	int nOfNodes = g->graphNodes.size();
	if (nOfNodes == 0) {
		cout << "zero nodes in the peace-generated input MST file!" << endl;
		exit(1);
	}

	//vector<vector<int> > nodes(nOfNodes-1, vector<int>(3));	//store edges in MST, there are n-1 edges, n is number of nodes.
	vector<vector<float> > nodes;

	//read mst from the input file
	ifstream in;
	in.open(mstFileName.c_str(), ios::in);

	if (in.good()) {
		string str;
		getline(in, str);

		while (str.size() != 0) {
			str.erase(str.find_last_not_of(" \n\r\t")+1); //trim str
			if (str[0] != '#') {	//comment line begins from '#'
				vector<string> paras = split(str, ',');
				if (atoi(paras[0].c_str()) != -1) {	//-1 means root of MST
					vector<float> tmp(3);
					tmp[0] = atof(paras[0].c_str());
					tmp[1] = atof(paras[1].c_str());
					tmp[2] = atof(paras[2].c_str());
					nodes.push_back(tmp);
				}
			}
			getline(in, str);
		}
	}
	in.close();

	// Make a undirected MST.
	DefGraph mst(nOfNodes);
	for (int j=0; j<nodes.size(); j++) {
		addEdge(mst, int(nodes[j][0]), int(nodes[j][1]), nodes[j][2], false);
	}

	g->setMst(mst);
}


vector<string> ESTAssembly::split(const string& str, char delimit) {
	vector<string> retStr;
	size_t found = str.find(delimit);
	if (found != string::npos) {
		int start = 0;
		while (found != string::npos) {
			retStr.push_back(str.substr(start, (int) found - start));
			start = found + 1;
			found = str.find(delimit, found + 1);
		}
		retStr.push_back(str.substr(start));
	}
	return retStr;
}

string ESTAssembly::toUpperCase(const string& str) {
	string retStr;
	for (int i=0; i<str.size(); i++) {
		retStr.push_back(toupper(str[i]));
	}
	return retStr;
}
